File: Assignment 5/C/05. insertion & deletion at a position in a doubly ll.c
#include <stdio.h>
#include <stdlib.h>

typedef struct node{
    struct node* prev;
    int val;
    struct node* next;
}node;

void traversal(node* head){
    node* ptr = head;
    printf("NULL <--");
    while(ptr){
        printf(" %d ",ptr->val);
        ptr = ptr->next;
        if(ptr){
            printf("<-->");
        }
    }
    printf("--> NULL\n");
}

node* insertion(node* head,int val,int pos){
    node* nn = (node*)malloc(sizeof(node));
    nn->val = val;
    nn->prev = NULL;
    nn->next = NULL;

    node* ptr = head;
    if(pos==1){
        ptr->prev = nn;
        nn->next = ptr;
        return nn;
    }
    for(int i=2; ptr->next!=NULL && i<pos; i++){
        ptr = ptr->next;
    }
    nn->next = ptr->next;
    nn->prev = ptr;
    ptr->next = nn;
    return head;
}

node* deletion(node* head,int pos){
    node* ptr = head;
    if(pos==1){
        ptr = head->next;
        ptr->prev = NULL;
        head->next = NULL;
        free(head);
        return ptr;
    }
    for(int i=2; ptr->next!=NULL && i<pos; i++){
        ptr = ptr->next;
    }
    node* temp = ptr->next;
    if (temp == NULL) { // if temp is the last node
        ptr->next = NULL;
    } else {
        ptr->next = temp->next;
        if (temp->next != NULL) {
            temp->next->prev = ptr;
        }
    }
    temp->next = NULL;
    temp->prev = NULL;
    free(temp);
    return head;
}

int main(){
    node* head = (node*)malloc(sizeof(node));
    node* snd = (node*)malloc(sizeof(node));
    node* trd = (node*)malloc(sizeof(node));
    node* four = (node*)malloc(sizeof(node));

    head->val = 1;
    snd->val = 2;
    trd->val = 3;
    four->val = 4;

    head->prev = NULL;
    head->next = snd;

    snd->prev = head;
    snd->next = trd;
    
    trd->prev = snd;
    trd->next = four;

    four->prev = trd;
    four->next = NULL;

    head = insertion(head,99,5);
    traversal(head);

    head = deletion(head,3);
    traversal(head);

    return 0;
}

File: Assignment 5/Python/05. insertion & deletion at a position in a doubly ll .py
class Node:
    def __init__(self, val):
        self.prev = None
        self.val = val
        self.next = None

def traversal(head):
    ptr = head
    print("NULL <--", end="")
    while ptr:
        print(f" {ptr.val} ", end="")
        ptr = ptr.next
        if ptr:
            print("<--> ", end="")
    print("--> NULL")

def insertion(head, val, pos):
    nn = Node(val)
    ptr = head
    if pos == 1:
        nn.next = ptr
        ptr.prev = nn
        return nn
    for _ in range(2, pos):
        ptr = ptr.next
    nn.next = ptr.next
    nn.prev = ptr
    ptr.next = nn
    return head

def deletion(head, pos):
    ptr = head
    if pos == 1:
        ptr = head.next
        ptr.prev = None
        head.next = None
        del head
        return ptr
    for _ in range(2, pos):
        ptr = ptr.next
    temp = ptr.next
    if temp is None:  # if temp is the last node
        ptr.next = None
    else:
        ptr.next = temp.next
        if temp.next is not None:
            temp.next.prev = ptr
    temp.next = None
    temp.prev = None
    del temp
    return head

head = Node(1)
snd = Node(2)
trd = Node(3)
four = Node(4)

head.prev = None
head.next = snd

snd.prev = head
snd.next = trd

trd.prev = snd
trd.next = four

four.prev = trd
four.next = None

head = insertion(head, 99, 5)
traversal(head)

head = deletion(head, 3)
traversal(head)



# OUTPUT:
# NULL <-- 1 <-->  2 <-->  3 <-->  4 <-->  99 --> NULL
# NULL <-- 1 <-->  2 <-->  4 <-->  99 --> NULL
